test/e2e_node/util.go

```diff
@@ -60,3 +70,155 @@ func getNodeSummary() (*stats.Summary, error) {
+	)
+
+	// Check whether a configmap for KubeletConfiguration already exists
+	_, err := getCurrentKubeletConfigMap(f)
+	if fmt.Sprintf("%v", err) == fmt.Sprintf("configmaps \"kubelet-%s\" not found", framework.TestContext.NodeName) {
+		_, err := createConfigMap(f, kubeCfg)
+		if err != nil {
+			return err
+		}
+	} else if err != nil {
+		return err
+	} else {
+		// The configmap exists, update it instead of creating it.
+		_, err := updateConfigMap(f, kubeCfg)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Wait for the Kubelet to restart.
+	time.Sleep(restartGap)
+
+	// Retrieve the new config and compare it to the one we attempted to set
+	newKubeCfg, err := getCurrentKubeletConfig()
+	if err != nil {
+		return err
+	}
+
+	// Return an error if the desired config is not in use by now
+	if !reflect.DeepEqual(*kubeCfg, *newKubeCfg) {
+		return fmt.Errorf("Either the Kubelet did not restart or it did not present the modified configuration via /configz after restarting.")
+	}
+	return nil
+}
+
+// Causes the test to fail, or returns a status 200 response from the /configz endpoint
+func pollConfigz(timeout time.Duration, pollInterval time.Duration) *http.Response {
+	endpoint := fmt.Sprintf("http://127.0.0.1:8080/api/v1/proxy/nodes/%s/configz", framework.TestContext.NodeName)
+	client := &http.Client{}
+	req, err := http.NewRequest("GET", endpoint, nil)
+	framework.ExpectNoError(err)
+	req.Header.Add("Accept", "application/json")
+
+	var resp *http.Response
+	Eventually(func() bool {
+		resp, err = client.Do(req)
+		if err != nil {
+			glog.Errorf("Failed to get /configz, retrying. Error: %v", err)
+			return false
+		}
+		if resp.StatusCode != 200 {
+			glog.Errorf("/configz response status not 200, retrying. Response was: %+v", resp)
+			return false
+		}
+		return true
+	}, timeout, pollInterval).Should(Equal(true))
+	return resp
+}
+
+// Decodes the http response  from /configz and returns a componentconfig.KubeletConfiguration (internal type).
+func decodeConfigz(resp *http.Response) (*componentconfig.KubeletConfiguration, error) {
+	// This hack because /configz reports the following structure:
+	// {"componentconfig": {the JSON representation of v1alpha1.KubeletConfiguration}}
+	type configzWrapper struct {
+		ComponentConfig v1alpha1.KubeletConfiguration `json:"componentconfig"`
+	}
+
+	configz := configzWrapper{}
+	kubeCfg := componentconfig.KubeletConfiguration{}
+
+	contentsBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	err = json.Unmarshal(contentsBytes, &configz)
+	if err != nil {
+		return nil, err
+	}
+
+	err = api.Scheme.Convert(&configz.ComponentConfig, &kubeCfg, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &kubeCfg, nil
+}
+
+// Uses KubeletConfiguration to create a `kubelet-<node-name>` ConfigMap in the "kube-system" namespace.
+func createConfigMap(f *framework.Framework, kubeCfg *componentconfig.KubeletConfiguration) (*api.ConfigMap, error) {
```

Comment:
Most code in `createConfigMap` and `updateConfigMap` is duplicated. Can we unify these 2 functions?